### DISCUSSION TITLE
announcement: sd-webui-model-mixer

### DISCUSSION THREAD

**Discussion by wkpark** [hako-mikan/sd-webui-supermerger#229] [Announcements]
This is another model merger/mixer.

It was created when I was looking for a way to use SuperMerger with other extensions. It doesn't have all the features of SuperMerger, but it inherits the core functionality.

## pros
- Supports merging multiple models sequentially (up to 5).
- Specialises in block-level merging to make block merging easier.
- Can be used without saving the merged model like SuperMerger.
- Supports saving already merged models.
- Merge information is saved in the image, and when the image is loaded, it will retrieve the settings used in the merged model.

## cons
- Doesn't support many of the calculation methods supported by SuperMerger.
- Does not support XYZ plots, etc. supported by supermerger.

I already have a PR for SuperMerger with hints from developing model-mixer,
I plan to add a PR to SuperMerger in the future to support as an extension.

https://github.com/wkpark/sd-webui-model-mixer

![image](https://github.com/hako-mikan/sd-webui-supermerger/assets/232347/0047228e-00eb-4754-9320-62f04e52016a)

**Comment by miasik** [comment]
Looks like it can process four models(or four mergings) in a line, right?
Please, add a display for the final formula for all the models enabled in a line.

**Comment by miasik** [comment]
How to select a different base model for add-diff for C, D, E?
Please, add an option to set Base model = Model A to add-diff as ModA+(ModB-ModA)

**Comment by miasik** [comment]
Please add an option to use first mergings in next steps. For example calculate ModC=ResultOfA+ResultOfB
Like here:
https://github.com/bbc-mc/sdweb-merge-board

**Comment by miasik** [comment]
Actually there is a Discussion section for the topics like this ;-)

**Reply by TomLucidor** [reply]
Kinda need help here on understanding KohyaSS (know it is not appropriate but dunno how to contact but similar ideas would apply to SuperMerger) https://github.com/bmaltais/kohya_ss/discussions/1751

**Comment by wkpark** [comment]
> Please add an option to use first mergings in next steps. For example calculate ModC=ResultOfA+ResultOfB Like here: https://github.com/bbc-mc/sdweb-merge-board

I've already tested that project, but it's too complicated.

The goal of ModelMixer is fast, quick, and easy testing.

> How to select a different base model for add-diff for C, D, E? Please, add an option to set Base model = Model A to add-diff as ModA+(ModB-ModA)

Currently, `ModA + (ModB- ModA)` could be rewritten as `Model A + (Model B - base_model)` and `base_model == ModA` case.
(Internally, all models are cached via webui's `checkpoint_info` and `checkpoints_loaded`)

All Add-difference operations assume that the `base_model` is the same. (e.g. `v1-5-pruned-emaonly.safetensors` for SD-v1.5 or `v2-1_768-nonema-pruned.safetensors` for SD-2.1 or `sd_xl_base_1.0.safetensors [31e35c80fc4]` for SDXL. if not specified, base-model will be auto-detected)

> Looks like it can process four models(or four mergings) in a line, right? Please, add a display for the final formula for all the models enabled in a line.

good suggestion. (will be added soon)

the number of models could be configured by webui settings, (The max models could be 2~5 and this limitation is only for practical reason.)

(currently, the final formula named "`Model recipe:`" will be written in the `png` "`parameters:`" information.)

**Reply by TomLucidor** [reply]
How would you advise people who want to make a "pack" with like 10+ LoRAs?

**Comment by miasik** [comment]
I wish I had SM as a node for ComfyUI.
I'd like to do multi-step mergings but saving and switching models is so boring

**Reply by TomLucidor** [reply]
Seconded this but nobody would do this, to have a LoRA/LyCORIS manufacturing tool

**Comment by wkpark** [comment]
> How would you advise people who want to make a "pack" with like 10+ LoRAs?

supermerger already can do this job.

and recently, Model Mixer does this job via sd-webui internal functionality with model-mixer adapted kyoha_ss script.

----

sd-webui itself has LoRA merger functionality internally,

you can test any LoRA/LyCORIS combinations with webui prompt easily,
- internally, all used LoRAs/LyCORISs are merged(weighted) into the currently selected checkpoint model.
- now, if you can save the current weighted/merged checkpoint model with LoRAs, named B, theoretically.
- `B model` - `Original model` --> this result could be converted into LoRA/LyCORIS.

SuperMerger internally does this job via lora merger (`pluslora.py` + adapted kyoha_ss script)

ModelMixer does this job via sd-webui internal function + adapted kyoha_ss script.
